packages/emnapi/src/emnapi.c

```diff
@@ -154,69 +157,163 @@ emnapi_get_emscripten_version(napi_env env,
 
 #ifdef __EMSCRIPTEN_PTHREADS__
 
+// #ifdef __wasm64__
+// #define __EMNAPI_ASYNC_SEND_CALLBACK_SIG \
+//   (EM_FUNC_SIG_RETURN_VALUE_V | \
+//   EM_FUNC_SIG_WITH_N_PARAMETERS(1) | \
+//   EM_FUNC_SIG_SET_PARAM(0, EM_FUNC_SIG_PARAM_I64))
+// #else
+// #define __EMNAPI_ASYNC_SEND_CALLBACK_SIG EM_FUNC_SIG_VI
+// #endif
+
+#ifndef EMNAPI_ASYNC_SEND_TYPE
+#define EMNAPI_ASYNC_SEND_TYPE 0
+#endif
+
+#if EMNAPI_ASYNC_SEND_TYPE == 0
+extern void _emnapi_set_immediate(void (*callback)(void*), void* data);
+#define NEXT_TICK(callback, data) _emnapi_set_immediate((callback), (data))
+#elif EMNAPI_ASYNC_SEND_TYPE == 1
+extern void _emnapi_next_tick(void (*callback)(void*), void* data);
+#define NEXT_TICK(callback, data) _emnapi_next_tick((callback), (data))
+#else
+#error "Invalid EMNAPI_ASYNC_SEND_TYPE"
+#endif
+
+extern void _emnapi_async_send_js(int type,
+                                  void (*callback)(void*),
+                                  void* data);
+
+static void _emnapi_async_send(void (*callback)(void*), void* data) {
+  // TODO(?): need help
+  // Neither emscripten_dispatch_to_thread_async nor MAIN_THREAD_ASYNC_EM_ASM
+  // invoke the async complete callback if there is a printf() in worker thread.
+  // This breaks "packages/test/pool" tests.
+  // Not sure what happens, maybe has deadlock,
+  // and not sure whether this is Emscripten bug or my incorrect usage.
+  // BTW emscripten_dispatch_to_thread_async seems
+  // not support __wasm64__ V_I64 signature yet
+
+  // pthread_t main_thread = emscripten_main_browser_thread_id();
+  // if (pthread_equal(main_thread, pthread_self())) {
+  //   NEXT_TICK(callback, data);
+  // } else {
+  //   emscripten_dispatch_to_thread_async(main_thread,
+  //                                       __EMNAPI_ASYNC_SEND_CALLBACK_SIG,
+  //                                       callback,
+  //                                       NULL,
+  //                                       data);
+  //   // or
+  //   // MAIN_THREAD_ASYNC_EM_ASM({
+  //   //   emnapiGetDynamicCalls.call_vp($0, $1);
+  //   // }, callback, data);
+  // }
```

Comment:
@RReverser  I left comments here. If you prefer that emnapi should not depend on `PThread` object, I would appreciate it if you could open another PR to provide help. Sorry for my broken English, forget it if there is some strange or offensive expression.

Reply:
@RReverser [c3a3cdb1](https://github.com/toyobayashi/emnapi/commit/c3a3cdb12c79fa4881758aa75bf7790bca166ab7) At last I found `emscripten_proxy_async` and `em_proxying_queue_create` works well, but using `emscripten_proxy_get_system_queue()` will also stuck if call printf in child thread. I could not figure out why.

Reply:
> If you prefer that emnapi should not depend on PThread object, I would appreciate it if you could open another PR to provide help

Heh unfortunately I'm finding a bit hard to follow how threading works in this library, so not sure I can submit a PR for that yet.

Reply:
FWIW one other reason I want it to have its own pool of workers is that upstream there are talks of making it possible to use pthreads without a pthread pool, and I now have a working implementation of that at least for Node.js.

However, it seems that emnapi won't work with that approach in the current form since it manipulates internal arrays and waits until `unusedWorkers` is not empty. But, if Emscripten will use pthreads without a pthread pool, it will be empty at the start so emnapi will wait forever.

Reply:
Ah, looks like I had a slightly outdated dist version and you already removed some more usages. This is looking promising, going to give it a try with the pool-less version of Emscripten.